### PR TITLE
Add deepseek3 conversion utility: orbax scan to hf 

### PR DESCRIPTION
**Pull request by shuningjin**
# Description
Onboard deepseek3-671b to checkpoint conversion utility, orbax scan -> hf

Other changes:
- `utils.py` - refactor of `process_leaf_param` function for clarity
-  `to_huggingface.py`- add timing and log.
-  `generate_hf_golden_logits.py` - add options to facilitate deepseek load

future work
- only works with local save, gcs save has max retries exceed: b/457821616
- hf -> orbax scan: mapping is there, need change stacking and optimize for to_maxtext, need test: b/457820372
- hf <-> orbax unscan: b/457820735

Fix: b/450671690

# Tests

Test details: b/450671690#comment12

## conversion
```
ID=$(date +%Y-%m-%d-%H-%M-%S)
RUN_NAME=ds3-hf-$ID

python3 -m MaxText.utils.ckpt_conversion.to_huggingface src/MaxText/configs/base.yml \
model_name=deepseek3-671b \
load_parameters_path=gs://ranran-multipod-dev/deepseek3/conversion/bf16/1/0/items \
base_output_directory=/home/shuningjin/deepseek3-671b/deepseek3-671b-hf-$ID \
scan_layers=true use_multimodal=false \
skip_jax_distributed_system=True attention=dot_product mla_naive_kvcache=false \
checkpoint_storage_concurrent_gb=1024 \
dtype=bfloat16 weight_dtype=bfloat16
```

- log: https://paste.googleplex.com/6229080404000768
- CPU, load orbax: 80min, transform: 50min, save to local: 440min

## forward logit check 

generate logits from newly generated hf checkpoint, compare it with the logit from orbax checkpoint, max KL=0.14

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py [hunk 1]
```diff
@@ -814,6 +814,122 @@ def reshape_kernel(input_tensor, target_shape):
...
+      "DeepSeekMoeBlock_0-MoeBlock_0-wo": "down_proj.weight",
+  }
+  for maxtext_key, hf_key in moe_expert_keys.items():
+    mapping[f"params-decoder-moe_layers-{maxtext_key}"] = [
+        [f"model.layers.{l}.mlp.experts.{e}.{hf_key}" for l in range(first_num_dense_layers, num_main_layers)]
+        for e in range(num_experts)
+    ]
+
+
+def DEEPSEEK_MAXTEXT_TO_HF_PARAM_HOOK_FN(config, scan_layers=False, saving_to_hf=True):
```
Comment (abhinavclemson) [inline src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: This Is adding  all layer for both cases dense and moe:

Could something like this works?
```
  if is_moe:
    # MoE Layers
    moe_layer_keys = attention_keys | {
        "DeepSeekMoeBlock_0-shared_experts-wi_0-kernel": "mlp.shared_experts.gate_proj.weight",
        "DeepSeekMoeBlock_0-shared_experts-wi_1-kernel": "mlp.shared_experts.up_proj.weight",
        "DeepSeekMoeBlock_0-shared_experts-wo-kernel": "mlp.shared_experts.down_proj.weight",
        "DeepSeekMoeBlock_0-MoeBlock_0-gate-kernel": "mlp.gate.weight",
        "DeepSeekMoeBlock_0-MoeBlock_0-gate-bias": "mlp.gate.e_score_correction_bias",
    }
    for maxtext_key, hf_key in moe_layer_keys.items():
      mapping[f"params-decoder-moe_layers-{maxtext_key}"] = [
          f"model.layers.{i}.{hf_key}" for i in range(first_num_dense_layers, num_main_layers)
      ]

    # MoE Experts (nested list mapping: [[e0_l0, e0_l1..], [e1_l0, e1_l1..]..])
    moe_expert_keys = {
        "DeepSeekMoeBlock_0-MoeBlock_0-wi_0": "gate_proj.weight",
        "DeepSeekMoeBlock_0-MoeBlock_0-wi_1": "up_proj.weight",
        "DeepSeekMoeBlock_0-MoeBlock_0-wo": "down_proj.weight",
    }
    for maxtext_key, hf_key in moe_expert_keys.items():
      mapping[f"params-decoder-moe_layers-{maxtext_key}"] = [
          [f"model.layers.{l}.mlp.experts.{e}.{hf_key}" for l in range(first_num_dense_layers, num_main_layers)]
          for e in range(num_experts)
      ]
  else:
    # Dense Layers
    dense_layer_keys = attention_keys | {
        "mlp-wi_0-kernel": "mlp.gate_proj.weight",
        "mlp-wi_1-kernel": "mlp.up_proj.weight",
        "mlp-wo-kernel": "mlp.down_proj.weight",
    }
    for maxtext_key, hf_key in dense_layer_keys.items():
      mapping[f"params-decoder-dense_layers-{maxtext_key}"] = [
          f"model.layers.{i}.{hf_key}" for i in range(first_num_dense_layers)
      ]
```

Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: Could you explain a bit? deepseek3-671b is having [61 layers](https://github.com/AI-Hypercomputer/maxtext/blob/d29bb51868e0ecf4964297ee06c3e0607e645281/src/MaxText/configs/models/deepseek3-671b.yml#L25-L26), with a mix of layers: 3 dense followed by 58 moe. correspondingly, I wrote
```
range(first_num_dense_layers)
range(first_num_dense_layers, num_main_layers)
```

File: src/MaxText/scratch_code/generate_hf_golden_logits.py [hunk 2]
```diff
@@ -110,7 +127,7 @@ def save_golden_logits(
     # 2. Run inference
     with torch.no_grad():
       outputs = model(**inputs)
-      logits = outputs.logits.cpu().numpy().astype("float32")
+      logits = outputs.logits.cpu().to(torch.float32).numpy()
```
Comment (RissyRan) [inline src/MaxText/scratch_code/generate_hf_golden_logits.py:130]: curious, is this due to the dependency change or performance impact?

File: src/MaxText/utils/ckpt_conversion/to_huggingface.py [hunk 3]
```diff
@@ -70,10 +70,12 @@
...
 from MaxText.utils.ckpt_conversion.utils.hf_model_configs import HF_MODEL_CONFIGS
 from MaxText.utils.ckpt_conversion.utils.utils import (process_leaf_param, save_model_files, HF_IDS)
+import time
+from tqdm import tqdm
 
 
-jax.config.update("jax_platform_name", "cpu")
-
+os.environ["JAX_PLATFORMS"] = "cpu"
+os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=16"
```
Comment (RissyRan) [inline src/MaxText/utils/ckpt_conversion/to_huggingface.py:78]: Glad to see it here! Did you see any improvement yet?

File: src/MaxText/utils/ckpt_conversion/utils/utils.py [hunk 4]
```diff
@@ -106,6 +119,7 @@ def process_leaf_param(
       return []  # Skip this parameter
 
   maxtext_param_key = "params-" + "-".join(key_parts)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/utils/utils.py:122]: Let's update maxtext.logging()

[the remark this answers is not in the document]
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/utils.py:122]: I want to track which maxtext parameter is being converted.

**Review by RissyRan — COMMENTED** [review]
Thank you! Could you help create 2 bugs to track the future work for DeepSeek you mentioned, and also has a TODO in the source code?

File: src/MaxText/utils/ckpt_conversion/utils/hf_model_configs.py [hunk 5]
```diff
@@ -469,6 +469,65 @@
...
+    "tie_word_embeddings": False,
+    "topk_group": 4,
+    "topk_method": "noaux_tc",
+    "torch_dtype": "bfloat16",
+    "transformers_version": "4.33.1",
+    "use_cache": True,
+    "v_head_dim": 128,
+    "vocab_size": 129280,
+}
+deepseek3_671b_config = transformers.DeepseekV3Config(**deepseek3_671b_dict)
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/utils/hf_model_configs.py:528]: Thanks and just out of curiosity, why not directly combine the above two parts into a single deepseek3_671b_config declaration?

File: src/MaxText/utils/ckpt_conversion/utils/hf_shape.py [hunk 6]
```diff
@@ -208,19 +208,158 @@ def GEMMA2_HF_WEIGHTS_TO_SHAPE(config):
   return mapping
 
 
+def DEEPSEEK_HF_WEIGHTS_TO_SHAPE(config):
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/utils/hf_shape.py:211]: Is it supporting both V2 and V3? If not, would you prefer `DEEPSEEK3` or  `DEEPSEEK` in the naming? Same applies for the name/hook mappings below.

File: src/MaxText/utils/ckpt_conversion/utils/utils.py [hunk 7]
```diff
@@ -87,6 +88,18 @@ def _get_local_directory(output_dir: str) -> str:
   return local_dir
 
 
+def _process(hf_path, processed_slice, output_weights, current_hook_fns, shape_map_local):
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/utils/utils.py:91]: Could you add a doc string to this function?

File: src/MaxText/utils/ckpt_conversion/to_huggingface.py [hunk 8]
```diff
@@ -114,17 +116,22 @@ def main(argv: Sequence[str]) -> None:
...
   os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
 
+  # Initialize maxtext config
   config = pyconfig.initialize(argv)
   assert (
       config.load_full_state_path == ""
   ), "This script expects parameters, not a full state. Use generate_param_only_checkpoint first if needed."
   max_utils.print_system_information()
 
+  # Load Maxtext checkpoint
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/to_huggingface.py:127]: +1 can we use `max_logging.log()`? Not urgent though

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py [hunk 9]
```diff
@@ -814,6 +814,122 @@ def reshape_kernel(input_tensor, target_shape):
...
+  moe_expert_keys = {
+      "DeepSeekMoeBlock_0-MoeBlock_0-wi_0": "gate_proj.weight",
+      "DeepSeekMoeBlock_0-MoeBlock_0-wi_1": "up_proj.weight",
+      "DeepSeekMoeBlock_0-MoeBlock_0-wo": "down_proj.weight",
+  }
+  for maxtext_key, hf_key in moe_expert_keys.items():
+    mapping[f"params-decoder-moe_layers-{maxtext_key}"] = [
+        [f"model.layers.{l}.mlp.experts.{e}.{hf_key}" for l in range(first_num_dense_layers, num_main_layers)]
+        for e in range(num_experts)
+    ]
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/utils/param_mapping.py:881]: Maybe I missed something, but this function should return a mapping right? `return mapping`

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py [hunk 10]
```diff
@@ -814,6 +814,122 @@ def reshape_kernel(input_tensor, target_shape):
...
+      "params-decoder-moe_layers-DeepSeekMoeBlock_0-shared_experts-wi_1-kernel",
+      "params-decoder-moe_layers-DeepSeekMoeBlock_0-shared_experts-wo-kernel",
+      "params-decoder-moe_layers-DeepSeekMoeBlock_0-MoeBlock_0-gate-kernel",
+      "params-decoder-moe_layers-DeepSeekMoeBlock_0-MoeBlock_0-wi_0",
+      "params-decoder-moe_layers-DeepSeekMoeBlock_0-MoeBlock_0-wi_1",
+      "params-decoder-moe_layers-DeepSeekMoeBlock_0-MoeBlock_0-wo",
+  }
+
+  for key in params_need_reshape:
+    mapping[key] = reshape_kernel
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/utils/param_mapping.py:934]: Same here, it will need to return a hook mapping right?

File: src/MaxText/utils/ckpt_conversion/to_huggingface.py (hunk 8 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/to_huggingface.py:127]: Thanks! updated to max_logging.log

File: src/MaxText/utils/ckpt_conversion/utils/hf_shape.py (hunk 6 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/hf_shape.py:211]: For V2, I hope to reuse this while adding if/else. 
- For example, I have added `if q_lora_rank is None`, which should work for V2, but haven't tested.

File: src/MaxText/scratch_code/generate_hf_golden_logits.py (hunk 2 above)
Comment (shuningjin) [reply src/MaxText/scratch_code/generate_hf_golden_logits.py:130]: This is because I load hf model with bfloat16.
- Previously, we are using `from_pretrained`  with dtype float32. This can take hours for deepseek3-671b.
- I add option to load with dtype bfloat16. The load time is 4min.
- As a result, logits is bfloat16, incompatible with numpy, have to convert to float32 first.

File: src/MaxText/utils/ckpt_conversion/to_huggingface.py (hunk 3 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/to_huggingface.py:78]: I am copying this from standalone conversion scripts:
https://github.com/AI-Hypercomputer/maxtext/blob/d29bb51868e0ecf4964297ee06c3e0607e645281/src/MaxText/convert_deepseek_family_ckpt.py#L722-L723

I haven't tested the performance difference with `os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=16"`.

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py (hunk 9 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/param_mapping.py:881]: Great catch, added! This was accidentally lost when solving merge conflict.

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py (hunk 10 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/param_mapping.py:934]: Great catch, added! This was accidentally lost when solving merge conflict.

File: src/MaxText/utils/ckpt_conversion/utils/utils.py (hunk 7 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/utils.py:91]: Thanks! I add a doc string.

File: src/MaxText/utils/ckpt_conversion/utils/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/utils.py:122]: Thanks! updated to `max_logging.log`

File: src/MaxText/utils/ckpt_conversion/utils/hf_model_configs.py (hunk 5 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/hf_model_configs.py:528]: I copy paste the config.json directly here. The other way seems need manually type every argument, can be error-prone. 
- Is there an easier way to declare DeepseekV3Config?
- If needed, I can change this in a followup PR.

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py (hunk 1 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: Discussed offline. deepseek (v3, v2) has the similar structure: a few dense + moe layers. no need to add `if is_moe`.

File: src/MaxText/utils/ckpt_conversion/utils/hf_model_configs.py (hunk 5 above)
Comment (hengtaoguo) [reply src/MaxText/utils/ckpt_conversion/utils/hf_model_configs.py:528]: Since it's just a style issue, I think it's OK to drop for now and fix as a follow-up if needed.

File: src/MaxText/utils/ckpt_conversion/utils/utils.py [hunk 11]
```diff
@@ -559,6 +572,7 @@ def upload_state_dict_to_gcs(state_dict: dict, gs_bucket_path: str):
       state_dict: A PyTorch model's state_dict.
       gs_bucket_path: GCS destination (e.g., "gs://my-bucket/models/model.pt").
   """
+  # TODO(shuningjin): max retries exceeded when uploading hf checkpoint for deepseek3-671b, b/457821616
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/utils/utils.py:575]: This is worth investigation. I've tried dumping a gemma3-4b safetensors locally and works fine for HF decoding, but never for such big model to GCS. We can track that in the bug. Thanks for catching!
